### PR TITLE
[Sprint: 44] XD-2730: Allow deployment properties as a file reference

### PR DESCRIPTION
**Pull request by ericbottard**

### CODE REVIEW

File: spring-xd-shell/src/main/java/org/springframework/xd/shell/command/DeploymentOptionKeys.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * Used to factor out names of the options that are shared by all deploy commands.
+ *
+ * @author Eric Bottard
+ */
+public class DeploymentOptionKeys {
+	static final String PROPERTIES_OPTION = "properties";
+
+	static final String PROPERTIES_FILE_OPTION = "propertiesFile";
```
Comment (garyrussell) [inline spring-xd-shell/src/main/java/org/springframework/xd/shell/command/DeploymentOptionKeys.java]: Perhaps `propertiesFiles` 

    --propertiesFiles=foo.properties,myCommon.properties

??

File: spring-xd-shell/src/main/java/org/springframework/xd/shell/command/JobCommands.java [hunk 2]
```diff
@@ -333,9 +341,29 @@ public String restartJobExecution(
...
+
+		int which = Assertions.atMostOneOf(PROPERTIES_OPTION, properties, PROPERTIES_FILE_OPTION, propertiesFile);
+		Map<String, String> propertiesToUse;
+		switch (which) {
+			case 0:
+				propertiesToUse = DeploymentPropertiesFormat.parseDeploymentProperties(properties);
+				break;
+			case 1:
+				Properties props = new Properties();
+				try (FileInputStream fis = new FileInputStream(propertiesFile)) {
```
Comment (garyrussell) [inline spring-xd-shell/src/main/java/org/springframework/xd/shell/command/JobCommands.java]: `new PatternMatchingResourceResolvert().getResource(propertiesFile).getInputStream()`   ??

Would allow `file:`, `classpath:`, `http:`.

**Comment by garyrussell** [comment]
Otherwise LGTM except a bunch of files need copyright updates.

**Comment by ericbottard** [comment]
Both suggestions are interesting BUT 
1. we'd lose TAB completion for Files if we were to accept any resource (a String)
2. not sure `classpath:` makes much sense in this case (would mean the CP of the shell, which the user is not supposed to edit. And if he were, this is pretty equivalent to `file:`). So the question is: is there really any value for non-File protocol here?

I think we could work something out at shell level to keep TAB completion and conversion for a `Foo[]`.

To recap, 
1. needs Shell support, which I think we should add
2. may not be a good idea after all?

What do you think?

File: spring-xd-shell/src/main/java/org/springframework/xd/shell/command/JobCommands.java [hunk 3]
```diff
@@ -333,9 +341,29 @@ public String restartJobExecution(
...
+					props.load(fis);
+				}
+				propertiesToUse = DeploymentPropertiesFormat.convert(props);
+				break;
+			default:
+				throw new AssertionError();
+		}
+
+		jobOperations().deploy(name, propertiesToUse);
+		return String.format("Deployed stream '%s'", name);
```
Comment (ericbottard) [inline spring-xd-shell/src/main/java/org/springframework/xd/shell/command/JobCommands.java]: Whoops, copy/paste, should be "job"

**Comment by ericbottard** [comment]
WRT copyright, what do you mean? Missing copyrights or need year update?
If the latter, shouldn't we do a full pass now and on Jan 1st?

**Comment by garyrussell** [comment]
No problem; they were just ideas; I am ok with deferring 1 and dropping 2.

Regarding copyright, no, it's not a blanket update; we update the copyright if a file is __changed__ in a calendar year; 2014 becomes 2014-2015; 2013-2014 becomes 2013-2015.

I generally don't bother if it's an automated updated (e.g. a class moves and causes a bunch of import changes) but any changes of substance in a class should be updated.

**Comment by ericbottard** [comment]
Rebased, updated copyrights.

**Comment by ericbottard** [comment]
Created https://jira.spring.io/browse/SHL-173 to track requirement for 1.

I updated the copyright headers, so should be ready to merge if no further comment

**Comment by garyrussell** [comment]
Squashed; merged as 550197a6b2488f6cab9cd1cf007779cbe858c3e0

**Comment by ilayaperumalg** [comment]
It looks like we need to have this merged into 1.1.x as well. Re-opening.

**Comment by ilayaperumalg** [comment]
Also, this change needs to be applied as well: c97ddba591a373c168e63a345b1c5897caa966ad

**Comment by ericbottard** [comment]
This is a new feature, I don't see why it should go into the maintenance branch.

**Comment by ilayaperumalg** [comment]
I see the JIRA https://jira.spring.io/browse/XD-2730 fix version says 1.1.x

**Comment by ilayaperumalg** [comment]
@ericbottard I agree, I think it should be 1.2 then.

**Comment by ericbottard** [comment]
I'll stick to what the JIRA God said, but seems odd to me

**Comment by ilayaperumalg** [comment]
Closing this as this is in fact was intended for 1.2
